test/datasets/amazonreviewpolarity_test.py

```diff
@@ -0,0 +1,82 @@
+import os.path
+import random
+import string
+import tarfile
+from collections import defaultdict
+
+from parameterized import parameterized
+from torchtext.datasets.amazonreviewpolarity import AmazonReviewPolarity
+
+from ..common.case_utils import TempDirMixin
+from ..common.torchtext_test_case import TorchtextTestCase
+
+
+def get_mock_dataset(root_dir):
```

Comment:
I wonder if it is technically feasible to Mock the HTTP link as well?

Reply:
I can look into this. I think it would be a matter of mocking the `GDriveReader` to return the tar-file

Reply:
Should be doable but you want to be careful how much you want to monkey patch things. When test fixture gets complicated, it becomes error-prone, especially because money patching is not what we do in library code thus we are not as familiar with the technique. And test fixtures are not tested, it's much harder to find bugs in fixtures.

Reply:
Just to be clear, my query was to patch the URL (like the proposal to patch the MD5) and not the code (GDriveReader). Meaning that the patched URL would work like a normal URL but instead would fetch the mocked dataset. Is this the same understanding you had @Nayef211 , @mthrok ? I am not sure though if this is doable or if there are still concerns doing this?

Reply:
Let me make sure I understand your suggestion correctly @parmeet. The reason we are patching the `MD5` hash is because the tests would actually fail without it. Patching the URL in this case doesn't make sense to me since the `on_disk_cache` datapipe just expects to verify whether the files that were downloaded by the `GDriveReader` datapipe exists locally. What would the benefit of patching the URL to return the `tar.gz` archive accomplish that wouldn't be done by the AmazonReviewPolarity dataset implementation?

Reply:
I think the logic we should be testing is:

1a. If the MD5 doesn't match, assert that we make an HTTP call
1a. If the MD5 does match, assert we read from cache
2a. If we've cached the contents of the tar, assert that we don't read the tar again
2b. If we haven't, assert we extract.
3. Test the parsing logic simply

Reply:
> I think the logic we should be testing is:
> 
> 1a. If the MD5 doesn't match, assert that we make an HTTP call 1a. If the MD5 does match, assert we read from cache 2a. If we've cached the contents of the tar, assert that we don't read the tar again 2b. If we haven't, assert we extract. 3. Test the parsing logic simply

I wonder if we have to test the logic for 1 and 2 since that's the expected behavior from the `OnDiskCacheHolderIterDataPipe` that we're using from torchdata. I assume their unit tests already do the check to verify that if the hash doesn't match, then retry the download and if the contents of an archive are cached, it shouldn't be extracted again. @ejguan would you be able to verify if this is indeed the case? Or do you think it's worthwhile adding tests for this as well?

I'm also following the same testing strategy used by torchaudio which assumes that any functions being used to implement the dataset has already been tested (in this case this would be datapipes). Instead we just test the API of the dataset to validate whether the expected input args result in the expected outputs. @mthrok could you weigh in with your thoughts as well?

Reply:
So as we are making the compressed file available in the folder, which is essentially the second step in the pipeline (the first being the download part). This means that any code/datapipe responsible for download part won't be executed/tested making it vulnerable for modifications.  The idea was simply to cover 100% of the code through testing :).

Reply:
> I'm also following the same testing strategy used by torchaudio which assumes that any functions being used to implement the dataset has already been tested (in this case this would be datapipes). 

I agree, we probably do not want to write tests that explicitly tests external library logic/workflow.

Reply:
`OnDiskCache` would check the hash, if it doesn't match, it will trigger the following ops (Downloader or Extract) in this case.

Reply:
> So as we are making the compressed file available in the folder, which is essentially the second step in the pipeline (the first being the download part). This means that any code/datapipe responsible for download part won't be executed/tested making it vulnerable for modifications. The idea was simply to cover 100% of the code through testing :).

BTW, as a very stupid example, if by mistake I end up changing GDriveReader to HttpReader, or change the url_dp to point to something else (instead of URL), I could completely bypass the current unit-testing.  That said, I don't want to block this PR for this. I don't mind merging it as such as it is already a huge improvement over what we had. Feel free to merge it without further due :)

Reply:
> 1a. If the MD5 doesn't match, assert that we make an HTTP call
> 1a. If the MD5 does match, assert we read from cache
> 2a. If we've cached the contents of the tar, assert that we don't read the tar again
> 2b. If we haven't, assert we extract.
> 3. Test the parsing logic simply

Regarding the @erip's points I think they are valid. As I do not know which component implements which logic, if 1 and 2 are what torchdata is responsible for and tested there, I do not think we need to test it here.

> I'm also following the same testing strategy used by torchaudio which assumes that any functions being used to implement the dataset has already been tested (in this case this would be datapipes). Instead we just test the API of the dataset to validate whether the expected input args result in the expected outputs. @mthrok could you weigh in with your thoughts as well?

In the context of mocking and patching, I prefer to avoid patching the code as much as possible, because in my experience, patching requires more knowledge about the implementation detail, and as the library code changes, these implementation detail changes, thus it increases the maintenance cost. Which is why I said "be careful how much you want to monkey patch stuff". (Of course, the notable exception to this is network call and such, which otherwise would not be possible.)
The reason why I chose to mock dataset is because they won't change. The datasets by nature are not supposed to change. So even when the library code changes dramatically, the implementation should be centered around the data. Thus we would not need to change the test fixture that much. (and as a bonus, we found a unexpected bug when the mock data was introduced. The order of files returned by glob was not consistent across OS. This would not be caught by patching)
Note that in torchaudio repo, we do not test the download part and the dataset implementation does not check cache if not downloading a new archive. so the code coverage is not 100%. I personally consider this a reasonable trade-off.

Reply:
Thanks for the detailed feedback and for describing how torchaudio approaches testing datasets @mthrok. Moving forward, I am going to stick to testing the core functionality of the dataset functions in this test suite. 

I still think @parmeet makes a valid point about getting better test coverage by patching `GDriveReader` datapipe to return the tar archive even if it adds a bit of maintenance overhead (like you said, its a tradeoff between maintenance and test coverage). I am going to merge this PR as is but let me add this patch in a followup PR so we can discuss further if it makes sense from a maintenance vs test coverage perspective.